tool/tsh/proxy.go

```diff
@@ -55,17 +55,110 @@ func onProxyCommandSSH(cf *CLIConf) error {
 		return trace.Wrap(err)
 	}
 
-	targetHost, targetPort, err := net.SplitHostPort(tc.Host)
+	proxyParams, err := getSSHProxyParams(cf, tc)
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	targetHost = cleanTargetHost(targetHost, tc.WebProxyHost(), tc.SiteName)
 
-	if tc.TLSRoutingEnabled {
-		return trace.Wrap(sshProxyWithTLSRouting(cf, tc, targetHost, targetPort))
+	if len(tc.JumpHosts) > 0 {
+		err := setupJumpHost(cf, tc, *proxyParams)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	if proxyParams.tlsRouting {
+		return trace.Wrap(sshProxyWithTLSRouting(cf, tc, *proxyParams))
 	}
 
-	return trace.Wrap(sshProxy(tc, targetHost, targetPort))
+	return trace.Wrap(sshProxy(tc, *proxyParams))
+}
+
+// getSSHProxyParams prepares parameters for establishing an SSH proxy.
+func getSSHProxyParams(cf *CLIConf, tc *libclient.TeleportClient) (*sshProxyParams, error) {
+	targetHost, targetPort, err := net.SplitHostPort(tc.Host)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// Without jump hosts, we will be connecting to the current Teleport client
+	// proxy the user is logged into.
+	if len(tc.JumpHosts) == 0 {
+		proxyHost, proxyPort := tc.SSHProxyHostPort()
+		if tc.TLSRoutingEnabled {
+			proxyHost, proxyPort = tc.WebProxyHostPort()
+		}
+		return &sshProxyParams{
+			proxyHost:   proxyHost,
+			proxyPort:   strconv.Itoa(proxyPort),
+			targetHost:  cleanTargetHost(targetHost, tc.WebProxyHost(), tc.SiteName),
+			targetPort:  targetPort,
+			clusterName: tc.SiteName,
+			tlsRouting:  tc.TLSRoutingEnabled,
+		}, nil
+	}
+	// When jump host is specified, we will be connecting to the jump host's
+	// proxy directly. Call its ping endpoint to figure out the cluster details
+	// such as cluster name, SSH proxy address, etc.
+	ping, err := webclient.Find(&webclient.Config{
+		Context:   cf.Context,
+		ProxyAddr: tc.JumpHosts[0].Addr.Addr,
```

Comment:
The JumpHost address provided by a user can point to "normal" (not TLSRouting multiplexed)  `SSHProxyListenPort == 3023` teleport  proxy ssh service. In that case this ping call (`webclient.Find(ProxyAddr: tc.JumpHosts[0].Addr.Addr`) will always fail.

Reply:
A user is supposed to point it to web proxy address. I updated the RFD to mention this as well in this PR. The SSH proxy address will be determined from the ping endpoint.